File: solutions/scala/zebra-puzzle/1/src/main/scala/Tester.scala
import Constraints.{HouseRule, NeighborRule}

object Tester {

    private def applies(person: Person, aspect: Any): Boolean =
        aspect match {
            case color: Color =>
                person.color == color
            case nationality: Nationality =>
                person.nationality == nationality
            case pet: Pet =>
                person.pet == pet
            case drink: Drink =>
                person.drink == drink
            case hobby: Hobby =>
                person.hobby == hobby
            case house: House =>
                person.house == house
            case _ =>
                person.house == aspect.asInstanceOf[House]
        }

    def test(
                person: Person,
                rule: HouseRule
            ): Boolean = {

        val subject = rule._1
        val fact    = rule._2
        val isTrue  = rule._3

        if applies(person, subject) then
            (isTrue && applies(person, fact)) || (!isTrue && !applies(person, fact))
        else
            (isTrue && !applies(person, fact)) || !isTrue
    }

    def test(
                set: Iterable[Person],
                rule: NeighborRule
            ): Boolean = {

        val persons = set.toArray
        val sorted  = persons.sortBy(_.house.ordinal)

        val pairs =
            sorted.zip(sorted.drop(1))

        rule._3 match {

            case Neighbor.LeftOf =>
                pairs.exists { case (left, right) =>
                    applies(left, rule._1) && applies(right, rule._2)
                }

            case Neighbor.RightOf =>
                pairs.exists { case (left, right) =>
                    applies(right, rule._1) && applies(left, rule._2)
                }

            case _ =>
                test(set, (rule._1, rule._2, Neighbor.LeftOf)) ||
                    test(set, (rule._1, rule._2, Neighbor.RightOf))
        }
    }
}

File: solutions/java/zebra-puzzle/1/src/main/java/Tester.java
import java.util.*;

final class Tester {

    static boolean applies(Person person, Object aspect) {
        if (aspect instanceof Color color) {
            return person.color() == color;
        }
        if (aspect instanceof Nationality nationality) {
            return person.nationality() == nationality;
        }
        if (aspect instanceof Pet pet) {
            return person.pet() == pet;
        }
        if (aspect instanceof Drink drink) {
            return person.drink() == drink;
        }
        if (aspect instanceof Hobby hobby) {
            return person.hobby() == hobby;
        }
        return person.house() == aspect;
    }

    static boolean test(Person person, HouseRule rule) {
        boolean subjectMatches = applies(person, rule.subject());
        boolean factMatches = applies(person, rule.fact());

        if (subjectMatches) {
            return rule.isTrue() == factMatches;
        } else {
            return !rule.isTrue()
                   || !factMatches;
        }
    }

    static boolean test(Set<Person> set, NeighborRule rule) {
        List<Person> sorted = new ArrayList<>(set);
        sorted.sort(Comparator.comparing(Person::house));

        for (int i = 0; i < sorted.size() - 1; i++) {
            Person left = sorted.get(i);
            Person right = sorted.get(i + 1);

            if (rule.relation() == Neighbor.LeftOf) {
                if (applies(left, rule.subject()) && applies(right, rule.fact())) {
                    return true;
                }
            }

            if (rule.relation() == Neighbor.RightOf) {
                if (applies(right, rule.subject()) && applies(left, rule.fact())) {
                    return true;
                }
            }
        }

        return rule.relation() == Neighbor.Either
               && isSatisfiedByAdjacentHouses(set, rule);
    }

    private static boolean isSatisfiedByAdjacentHouses(final Set<Person> set, final NeighborRule rule) {
        return test(set, new NeighborRule(rule.subject(), rule.fact(), Neighbor.LeftOf))
               || test(set, new NeighborRule(rule.subject(), rule.fact(), Neighbor.RightOf));
    }
}